### PR TITLE
Ongoing HTTP/2 progress

### PR DESCRIPTION
**Pull request by jrudolph**

### CODE REVIEW

**Comment by jrudolph** [comment]
Pushed lots of commits. Little actual implementation progress today but lots of progress on the testing infrastructure and actual tests.

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/rendering/HttpResponseHeaderHpackCompression.scala [hunk 1]
```diff
@@ -23,46 +24,28 @@ final class HttpResponseHeaderHpackCompression extends GraphStage[FlowShape[Http
...
 
   override def shape = FlowShape.of(in, out)
 
-  // format: OFF
-  override def createLogic(inheritedAttributes: Attributes) = new GraphStageLogic(shape)
-    with InHandler with OutHandler {
-    // format: ON
-
-    val buf = {
-      val b = ByteBuffer.allocate(4 * maxHeaderSize) // FIXME, just guessed a number here
```
Comment (jrudolph) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/rendering/HttpResponseHeaderHpackCompression.scala]: I think a reusable `ByteArrayOutputStream` should equally effective but less complex. WDYT?

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/HPackSpecExamples.scala [hunk 2]
```diff
@@ -43,6 +46,25 @@ object HPackSpecExamples {
...
+   *  Hex dump of encoded data:
+   *
+   *  8287 85bf 4088 25a8 49e9 5ba9 7d7f 8925 | ....@.%.I.[.}..%
+   *  a849 e95b b8e8 b4bf                     | .I.[....
+   */
+  val C43ThirdRequestWithHuffman =
+    """8287 85bf 4088 25a8 49e9 5ba9 7d7f 8925
+       a849 e95b b8e8 b4bf""".parseHexByteString
+
+  /**
```
Comment (ktoso) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/HPackSpecExamples.scala]: nice cleanup and addition

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala [hunk 3]
```diff
@@ -75,10 +75,11 @@ class Http2ServerDemux extends GraphStage[BidiShape[Http2SubStream, FrameEvent,
 
   def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
     new GraphStageLogic(shape) with BufferedOutletSupport { logic ⇒
-      class SubStream(
+      case class SubStream(
```
Comment (ktoso) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala]: I'm somewhat allergic to keeping `case class` not final unless there's a reason :)

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala [hunk 4]
```diff
@@ -109,6 +110,10 @@ class Http2ServerDemux extends GraphStage[BidiShape[Http2SubStream, FrameEvent,
 
               dispatchSubstream(sub)
 
+            case RstStreamFrame(streamId, errorCode) ⇒
+              // FIXME: also need to handle the other case when no response has been produced yet (inlet still None)
+              incomingStreams(streamId).inlet.foreach(_.cancel())
```
Comment (ktoso) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala]: to confirm my understanding, the cancel only is for a given stream and the connection stays alive... Did not yet see a test for that, reviewing further, would be good to add if not covered yet

> The RST_STREAM frame fully terminates the referenced stream and causes it to enter the "closed" state. After receiving a RST_STREAM on a stream, the receiver MUST NOT send additional frames for that stream, with the exception of PRIORITY. However, after sending the RST_STREAM, the sending endpoint MUST be prepared to receive and process additional frames sent on the stream that might have been sent by the peer prior to the arrival of the RST_STREAM.

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala [hunk 5]
```diff
@@ -299,18 +312,19 @@ class Http2ServerSpec extends AkkaSpec with WithInPendingUntilFixed {
   abstract class TestSetupWithoutHandshake {
     implicit def ec = system.dispatcher
 
-    val netIn = ByteStringSinkProbe()
-    val netOut = TestPublisher.probe[ByteString]()
+    val toNet = ByteStringSinkProbe()
+    val fromNet = TestPublisher.probe[ByteString]()
```
Comment (ktoso) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala]: nice rename

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/rendering/HttpResponseHeaderHpackCompression.scala [hunk 6]
```diff
@@ -23,46 +24,28 @@ final class HttpResponseHeaderHpackCompression extends GraphStage[FlowShape[Http
...
-      // feed `buf` with compressed header data
-      val headerBlockFragment = encodeAllHeaders(response)
+      val headerBlockFragment = encodeResponse(response)
 
-      // FIXME: emit proper exception if header is missing
-      val streamId = response.header[Http2StreamIdHeader].get.streamId
+      def failBecauseOfMissingHeader: Nothing =
+        // header is missing, shutting down because we will most likely otherwise miss a response and leak a substream
+        // TODO: optionally a less drastic measure would be only resetting all the active substreams
+        throw new RuntimeException("Received response for HTTP/2 request without Http2StreamIdHeader. Failing connection.")
```
Comment (ktoso) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/rendering/HttpResponseHeaderHpackCompression.scala]: while we address this todo, should introduce proper exception.
I think aborting like this is OK until proven too aggresive

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/rendering/HttpResponseHeaderHpackCompression.scala [hunk 7]
```diff
@@ -85,29 +68,18 @@ final class HttpResponseHeaderHpackCompression extends GraphStage[FlowShape[Http
           encoder.encodeHeader(os, nameBytes, valueBytes, false)
         }
 
-      // copy buffer to ByteString
-      mkByteString(buf)
+      val res = ByteString(os.toByteArray)
+      os.reset()
```
Comment (ktoso) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/rendering/HttpResponseHeaderHpackCompression.scala]: yeah, let's just use the `os` here, thanks for noticing; we'll see if it matters perf wise thtough

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/rendering/HttpResponseHeaderHpackCompression.scala [hunk 8]
```diff
@@ -85,29 +68,18 @@ final class HttpResponseHeaderHpackCompression extends GraphStage[FlowShape[Http
...
 
     setHandlers(in, out, this)
   }
 
 }
 
 object HttpResponseHeaderHpackCompression {
-  final val AuthorityKey = ":authority".getBytes
   final val StatusKey = ":status".getBytes
-  final val MethofKey = ":method".getBytes
```
Comment (ktoso) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/rendering/HttpResponseHeaderHpackCompression.scala]: ok

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala [hunk 9]
```diff
@@ -377,7 +420,7 @@ class Http2ServerSpec extends AkkaSpec with WithInPendingUntilFixed {
       encode(":method", request.method.value)
       encode(":scheme", request.uri.scheme.toString)
       encode(":path", request.uri.path.toString)
-      encode(":authority", request.uri.authority.toString)
+      encode(":authority", request.uri.authority.toString.drop(2) /* Authority.toString prefixes two slashes */ )
```
Comment (ktoso) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala]: thanks for the comment

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala [hunk 10]
```diff
@@ -159,9 +159,49 @@ class Http2ServerSpec extends AkkaSpec with WithInPendingUntilFixed {
...
+      "fail entity stream if peer sends RST_STREAM frame" inPendingUntilFixed new WaitingForRequestData {
+        val data1 = ByteString("abcdef")
+        sendDATA(1, endStream = false, data1)
+        entityDataIn.expectBytes(data1)
+
+        sendRST_STREAM(1, ErrorCode.INTERNAL_ERROR)
+        val error = entityDataIn.expectError()
+        error.getMessage should contain("Peer canceled stream with INTERNAL_ERROR(0x2) error code.")
+      }
+      "send RST_STREAM if entity stream is canceled" inPendingUntilFixed new WaitingForRequestData {
```
Comment (ktoso) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala]: OK, that's the test I was looking for

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/rendering/HttpResponseHeaderHpackCompression.scala (hunk 6 above)
Comment (jrudolph) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/rendering/HttpResponseHeaderHpackCompression.scala]: :+1:

**Pull request merged**